### DISCUSSION TITLE
Standardizing logging configuration in Dagster

### DISCUSSION THREAD

**Discussion by rexledesma** [dagster-io/dagster#12495] [Ideas]
# Overview
Logging is central to Dagster's core infrastructure, providing observability to computations defined using the framework as well as to Dagster components that aid in executing the computations.

Here are some examples where logging is used in the system:

1. [Capturing Python logs](https://docs.dagster.io/concepts/logging/python-logging) to send to the Dagster event log, allowing them to displayed in the default view of Dagit.
2. Displaying raw `stdout` and `stderr` logs using [Dagster's compute log storage](https://docs.dagster.io/deployment/dagster-instance#compute-log-storage).
3. Providing observability in Dagster infrastructure components, such as:
    1. The [Dagster server](https://docs.dagster.io/concepts/dagit/dagit) (the `uvicorn` server powering the Dagster UI and GraphQL interface),
    2. The [Dagster daemon](https://docs.dagster.io/deployment/dagster-daemon#dagster-daemon) (a long-running process that support features like scheduling, sensors, and run queueing).
    3. The [Dagster code locations](https://docs.dagster.io/concepts/code-locations) (powered by `gRPC`) that host the user's Dagster definitions and interface with the Dagster UI and the daemon.

# Problems

## Existing User Issues

Here is a sample of user-reported issues regarding our logging configuration, mostly dealing with (3) providing observability in Dagster infrastructure components. Currently, configuring logging in these components is not well supported. 

- https://github.com/dagster-io/dagster/issues/4170
- https://github.com/dagster-io/dagster/issues/4299
- https://github.com/dagster-io/dagster/issues/4992
- https://github.com/dagster-io/dagster/issues/6770
- https://github.com/dagster-io/dagster/issues/8911

# Proposed Solution

Standardization of Dagster's logging interface must occur. Here are a few high-level changes that need to happen:

- The names and default configuration of existing loggers should be documented.
- Expose a standardized interface to configure how logs are displayed, as well as what level of logs are displayed while maintaining the required server handlers to correctly route logs to the Dagster event logs.
    - This should follow best practices for Python frameworks that expose logging configuration. Ideally, something like [`uvicorn`'s `--log-config` setting](https://www.uvicorn.org/settings/#logging), which takes in a user-defined object.
    - This interface should be exposed for each Dagster infrastructure component (e.g. the Dagster server, daemon, and code locations)

**Comment by markfickett** [comment]
Thanks for starting this discussion!

Another area I'd love to see included is the ability to better send logs from Dagster to additional handlers, and filter them along the way. I think exposing a more standard interface to logging configuration meshes with this. Existing issues:

*   [Preserve original LogRecord when logging to custom handler](https://github.com/dagster-io/dagster/issues/7642)
*   [Preserve module/lineno/filename when sending log messages to @loggers](https://github.com/dagster-io/dagster/issues/7890)

And, some improved scalability with `context.log` usage in bit fan-outs with `DynamicOutput`, both for the events API and viewing logs in Dagit: [Improve performance in Dagit with large log volumes](https://github.com/dagster-io/dagster/issues/7821) .

**Reply by rexledesma** [reply]
Hi @markfickett, thanks for your input. Indeed, custom handlers are definitely under the scope of this standardization: the ideal state is that they should:

1. Be configurable as specified by the Python logging documentation, and
2. Not have any interference with the internal Dagster logging machinery.

To be clear, are you talking about logs emitted from ops/assets in Dagster, using the `context` or user-specified loggers? Again, the framework will still inject its own handler to ensure that the specified logs are emitted to the Dagster event logs, but custom handlers should be injected in such a way that it processing logs that are free of any Dagster pre-processing.

**Reply by markfickett** [reply]
I'm interested in both.

One use case is customizing the Dagster console output. I wanted to adjust the formatting, as well as filter out some spcific noisy logs. So I want that handler to include both `context.log` calls I'm making, other Python logger calls in my application code, and logs from internal Dagster loggers.

Another use case is sending logs to a structured JSON file for fluentbit to pick up. Again, I want to include all of Dagster-framework ogs, `context.log` messages, and other Python loggers from my application.

So, if I'm able to use the standard Python logging framework to specify a custom Handler, as well as its filters, that sounds great!

I had been using `managed_python_loggers` to make sure all logs go to my handler, and `python_log_level` to exclude some Dagster framework debug logs.

**Comment by tha23rd** [comment]
Just throwing a comment in here to advocate for an enhancement on customizing logging in Dagster internals. Our use case is that we would like to export logs created by Daemons to Datadog for monitoring + alerting purposes. At a bare minimum it would be a huge win if we could simply export the logs in JSON format.

Thanks!

**Comment by sigwinch28** [comment]
There have been several releases since this discussion and since the related issues have last been updated. Could we get a progress/planning update on this, please?

**Comment by sean-stach-ff** [comment]
Any update on this? So far the only solution we have found is to manually set a custom logger for all our jobs.

**Comment by dorothychen** [comment]
Would also love to understand if there are planning updates here.

To describe our use case - we also use DataDog for monitoring, and the inability to globally override how Dagster system logs (in particular, error log stack traces) are formatted has been a big usability pain point for developers. 

(I've spoken to some folks on the Dagster side about this on slack as well, but I want to tie that together with this more centralized thread)

**Reply by dorothychen** [reply]
From my (admittedly limited) perspective given our specific use case, some potential solutions that would cover the gap for us:

- if there was a way to indicate that a [custom python handler](https://docs.dagster.io/concepts/logging/python-logging#configuring-python-log-handlers-) should replace (instead of add to) the default dagster logger
- if there was a way to set a custom logger for all jobs in a code location by default, without needing to be specified in each job's runconfig
- if there was a way to customize/inject a default system logger directly, i.e. replace `colored_console_logger` [here](https://github.com/dagster-io/dagster/blob/e4e166b98c169e8ca4dd0f223a47199fbf453c70/python_modules/dagster/dagster/_loggers/__init__.py#L104)

**Reply by sean-stach-ff** [reply]
@dorothychen I was able to apply custom logger to the entire dagster instance using https://docs.dagster.io/concepts/logging/python-logging#configuring-python-log-handlers-

**Reply by dorothychen** [reply]
Have you been able to get the custom logger to replace dagster's system loggers, i.e. for error messages and stack traces? This is the mechanism I'm referring to in my first bullet point (had the wrong link originally; updated!), but unfortunately afaict using the custom logger is only additive ([code pointer](https://github.com/dagster-io/dagster/blob/dc407105d25869e8c506b2f15ce378ca44f0d31d/python_modules/dagster/dagster/_core/instance/__init__.py#L2358))

**Reply by sean-stach-ff** [reply]
My problem was that I needed all logs to be formatted in a way to be ingested. I used the above linked concept to add a new formatter which formats the log and outputs them to STDOUT.

I suppose this doesn't help replace the logs that are already in place but it does allow for a global custom logger without having to set it each time. However, since the logs are appearing in STDOUT and Dagster default logs appear in STDERR - it does have a clear separation.

**Reply by mdwcrft** [reply]
Did you solve the duplicate logger issue in the end @dorothychen? Having exactly the same problem now with Datadog

**Reply by mdwcrft** [reply]
This config worked for me as a workaround:
```
pythonLogs:
  pythonLogLevel: CRITICAL # To filter out the default logger
  managedPythonLoggers:
    - dagster
    - pipelines
  dagsterHandlerConfig:
    handlers:
      json_stderr:
        class: logging.StreamHandler
        formatter: json
        level: NOTSET
        stream: ext://sys.stdout
    formatters:
      json:
        '()': dagster.JsonLogFormatter
```

**Comment by zyd14** [comment]
Any updates here? Logging with anything except the default loggers is still a huge pain, especially if you want global configuration. Also I'm not seeing an obvious way to pass loggers for assets and asset jobs - there's logger_defs for op-jobs, but I don't see anything similar for assets / asset jobs. Has anyone figured out how to provide custom loggers for assets?

**Comment by VerdantForge** [comment]
So far I've been using this hack to be able to log to dagster console and use our own custom logger (which forwards logs to an aggregator)

```python
class MultiLogger(Logger):
    """Log to Dagster and to the default logger"""

    python_logger: Logger
    dagster_logger: Logger

    def __init__(
        self,
        name,
        python_logger: Logger,
        dagster_logger: Logger,
        level=settings.LOG.LEVEL,
    ):
        super().__init__(name, level=level)
        self.python_logger = python_logger
        self.dagster_logger = dagster_logger

    def _log(self, level, msg, args, **kwargs):
        self.python_logger.log(level, msg, *args, **kwargs)
        self.dagster_logger.log(level, msg, *args, **kwargs)


def get_dagster_and_python_logger(name: str) -> Logger:
    return MultiLogger(
        name,
        python_logger=logging.getLogger(name),
        dagster_logger=get_dagster_logger(name),
    )
```

**Reply by zyd14** [reply]
So then you just call `get_dagster_and_python_logger` to get a logger instead of using `context.log`? Unfortunately not really feasible for me as I'd have to convert a couple hundred ops / assets to use this instead of `context.log`.

**Reply by VerdantForge** [reply]
Usually a call to `logger = get_dagster_and_python_logger(__name__)` at the top of the module to get a module level logger, the methods then use `logger.info`, `logger.error`, etc...
A quick regex should be able to take care of your refactoring, no?

**Comment by v1gnesh** [comment]
Any thoughts on using [eliot](https://pythonspeed.com/articles/logging-for-scientific-computing/) for logging?

**Comment by theholy7** [comment]
Any updates on this?

**Comment by ssillaots-boku** [comment]
Any updates? Has someone been able to configure JSON logging to be the default format for all Dagster logging?
